### DISCUSSION TITLE
[Q&A] Does the FL Simulator use a GPU by default?

### DISCUSSION THREAD

**Discussion by apandit-algo** [NVIDIA/NVFlare#2227] [Q&A] [answered]
### Python version (`python3 -V`)

3.9.18

### NVFlare version (`python3 -m pip list | grep "nvflare"`)

2.3.8

### NVFlare branch (if running examples, please use the branch that corresponds to the NVFlare version, `git branch`)

2.3.8

### Operating system

Ubuntu 22.04

### Have you successfully run any of the following examples?

- [X] hello-numpy-sag with simulator
- [X] hello-pt with simulator
- [X] hello-numpy-sag with POC
- [X] hello-pt with POC

### Please describe your question

I am running the '[cifar10-sim](https://github.com/NVIDIA/NVFlare/tree/2.3.8/examples/advanced/cifar10/cifar10-sim)' advanced example using the FL Simulator command,
```nvflare simulator <job_path> -w <workspace_path> -n 2 -t 2```

I also have `nvidia-smi` utility installed on my machine.

My question is, 
1. If I do not mention the `-gpu` argument, does the simulator still run on a GPU by default instead of a CPU?
2. If so, how do I force the simulator to utilize a CPU despite having a GPU on my machine?

I read the documentation on [FL Simulator](https://nvflare.readthedocs.io/en/2.3/user_guide/fl_simulator.html), but am still a bit confused.

**Comment by chesterxgchen** [comment]
@yhwen  can you help answer this ?

**Reply by yhwen** [reply, marked as answer]
Sorry for missing this earlier and delay the response.

Simulator does not force the use GPU by default. When providing the "-gpu" option, each client running process will add the CUDA_VISIBLE_DEVICES environment variable and setting the process GPU group accordingly. If not providing the "-gpu" option, the client running process will directly execute the training job without adding the CUDA_VISIBLE_DEVICES variable. However, if you process session already have the CUDA_VISIBLE_DEVICES environment variable, it will inherit from it to run.

**Reply by apandit-algo** [reply]
Thanks for the clarification!

**Comment by chesterxgchen** [comment]
sorry for the delayed response.
